rfcs/2020-09-28-cloudwatch-metrics.md

```diff
@@ -0,0 +1,179 @@
+  request.rate_limit_num = 25 # optional, default
+  request.retry_attempts = 18446744073709551615 # optional, default
+  request.retry_initial_backoff_secs = 1 # optional, default, seconds
+  request.retry_max_duration_secs = 10 # optional, default, seconds
+  request.timeout_secs = 30 # optional, default, seconds
+
+  [[sources.cloudwatch.metrics]]
+    namespace = "AWS/EC2" # required
+    names = ["EBSReadOps", "EBSReadBytes", "Network*"] # optional; defaults to all metrics in namespace, ["*"], (refreshed on interval); supports globbing
+    dimensions.InstanceId = "i-05517fbc2e6124dfb" # optional; supported dimensions differ by namespace and metric; supports globbing
+    statistics = [ "average", "sum", "minimum", "maximum", "sample_count" ] # statistics to collect; can also contain extended statistics like p99; default shown
+    period = 300 # period (s) to aggregate metrics over; defaults to top-level `period` setting; top-level interval should be a multiple of this and any other defined periods
+```
+
+We could alternatively model dimensions as another table:
+
+```toml
+[[sources.cloudwatch.metrics.dimensions]]
+  key = "InstanceId"
+  value = "i-05517fbc2e6124dfb"
+```
+
+To support:
+
+* globbing metric names
+* fetching all metrics for a namespace
+* globbing of dimension values
+
+We will refresh and cache the available metrics on an interval
+(`metrics_refresh_interval`) for namespaces appearing in the configuration.
+
+A wildcard dimension value (`"*"`) will pull and publish discrete metrics for
+each dimension value. This is different than omitting the dimension altogether
+which would instead aggregate across the values for the dimension.
+
+Note that dimensions must match the dimensions that the metrics were published
+with. For example, if the metric was published two dimensions,
+availability-zone and load balancer name, then both dimensions must be
+specified to retrieve the metric. CloudWatch won't automatically aggregate
+across the other if only one is specified.
+
+This source will use the default AWS credential chain similar to the
+`aws_cloudwatch_metrics` sink.
+
+The output (shown here in prometheus format) will publish metrics in the form
+of:
+
+```
+<namespace>_<metric_name>_<metric_stat>{<dimension_key>=<dimension_value>*}
+```
+
+All metrics will be additionally tagged with the region.
+
+For example:
+
+```
+aws_ec2_cpu_credit_balance_average{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 576
+aws_ec2_cpu_credit_balance_maximum{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 576
+aws_ec2_cpu_credit_balance_minimum{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 576
+aws_ec2_cpu_credit_balance_sample_count{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 1
+aws_ec2_cpu_credit_balance_sum{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 576
+```
+
+I think we will want to provide example configs that collect common metrics for
+various namespaces (like EC2, S3, etc.).
```

Comment:
Would this be better solved with macros 😄 ?

Reply:
It could! :smile: